src/security/CertGadgets.h

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (C) 1996-2019 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_SRC_SECURITY_CERTGADGETS_H
+#define SQUID_SRC_SECURITY_CERTGADGETS_H
+
+#include "security/forward.h"
+
+namespace Security
+{
+
+/// \return the SubjectName field of the given certificate
+SBuf CertSubjectName(const CertPointer &);
+
+/// Squid defined error code (<0), an error code returned by X.509 API, or SSL_ERROR_NONE
+typedef int ErrorCode;
+
+/// \return true if cert was issued by the given issuer CA.
+///         sets checkCode to the library specific test result.
+bool CertIssuerCheck(const CertPointer &cert, const CertPointer &issuer, ErrorCode &checkCode);
```

Comment:
Whether this is an error is caller dependent. Some callers it is an error, some it is just a condition for different handling. 

Throwing from this function will guarantee that any OpenSSL enabled Squid builds using chains will fail to load on next restart/reconfigure. Since their cert is not self-signed and *will* be checked against itself at least once _within_ that chain load loop.

Reply:
> Whether this is an error is caller dependent

Error _handling_ may be caller dependent, but I see no caller-dependent error _detection_ here. Caller-dependent error handling is compatible with exceptions, of course.

> Throwing from this function will guarantee that any OpenSSL enabled Squid builds using chains will fail to load on next restart/reconfigure.

In general: When I suggest to do X, please assume that the suggestion includes whatever post-X adjustments are necessary.

In this particular case, if an OpenSSL build needs to handle errors/exceptions specially, then please make sure it does. However, I think you are misinterpreting the definition of an error here. "The given `issuer` did not sign the given `cert`" is _not_ an error as far as this function is concerned. The function should continue to return a boolean result and, naturally, that result may be false.

Reply:
> Please report errors using C++ exceptions, not error codes.

I took this to mean you wanted the "ErrorCode &checkCode" to be returned to the caller via exception.

Reply:
> I took this to mean you wanted the "ErrorCode &checkCode" to be returned to the caller via exception.

As I said, I would start by throwing TextException. This implies _not_ returning library-specific error codes to the caller. IIRC, the current callers do not care about library-specific error codes. And I expect future callers to remain ignorant about them. Callers just want to know whether the issuer signed the certificate. The function should return that information to the caller. If the function cannot tell, it should throw an exception. This is basic C++ programming, nothing fancy going on here!

And, again, incorporating library-specific error text into that exception text is probably a good idea.

Reply:
I have attempted this and it spreads SBuf dependencies around a lot of code that should not be necessary to touch by this PR and a few #define collisions in helpers (which also gain libsbuf dependency).

The caller logic also becomes quite weird in places. Like so:
  try {
    CertSelfSignedCheck(cert);
    // this is the error handling code path
    return;
  } catch (...) {
    // this is the desired normal code path
  }

Overall it looks to me like returning a bool where both true and false have meaning and an optional errorCode the caller MAY (or not) use is the cleaner change.

Reply:
The [example](https://github.com/squid-cache/squid/pull/458#discussion_r352557996) does look weird indeed. It also does not match my suggestions. I will try to find the time to provide the API changes I was talking about. Do you prefer that I create a new PR (against this PR branch; in your Squid repository) or should I modify your branch (and this PR) directly?

Reply:
A new branch based on the one for this PR should do - I just need to see what the proposal would look like and merge it if agreed. If you want to go so far as a PR against this PR's branch then fine.

Reply:
Please see https://github.com/measurement-factory/squid/tree/yadij-v5-gnutls-chainload-extra branch and its commit https://github.com/measurement-factory/squid/commit/a8c8d9e3dbeb8886eb60b35d5fb35f8ce62fe730. I believe that commit would address this change request in principle, but I cannot compile code right now, so it is not tested at all and probably needs some further polishing.

Reply:
Merged, thank you.